### PR TITLE
Guard against non-public header inclusion

### LINKED ISSUE #4856: Revise Public and Private Header Files in Kokkos
This issue tracks revision of public vs. private headers in Kokkos.  Users can include public headers, but not private.  We are revising the public and private headers design because ... ??

### PR DESCRIPTION
**Pull request by crtrott**
Fixes #4856.

### CODE REVIEW

**Comment by masterleinad** [comment]
Looks like the indentation is still off.

File: core/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -69,6 +69,8 @@ KOKKOS_INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR})
 KOKKOS_INCLUDE_DIRECTORIES(REQUIRED_DURING_INSTALLATION_TESTING ${CMAKE_CURRENT_SOURCE_DIR})
 KOKKOS_INCLUDE_DIRECTORIES(${KOKKOS_SOURCE_DIR}/core/unit_test/category_files)
 
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DKOKKOS_IMPL_PUBLIC_INCLUDE")
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/unit_test/CMakeLists.txt:72]: so you want me to do that for each target in this file?

File: core/src/KokkosExp_InterOp.hpp [hunk 2]
```diff
@@ -42,6 +42,15 @@
 //@HEADER
 */
 
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
+#include <Kokkos_Macros.hpp>
+#ifndef KOKKOS_ENABLE_DEPRECATED_CODE_3
+static_assert(false,
+              "Including non-public Kokkos header files is not allowed.");
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/KokkosExp_InterOp.hpp]: made it public

File: core/src/Kokkos_Graph_fwd.hpp [hunk 3]
```diff
@@ -42,6 +42,15 @@
 //@HEADER
 */
 
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/Kokkos_Graph_fwd.hpp]: made it public

File: core/src/KokkosExp_MDRangePolicy.hpp [hunk 4]
```diff
@@ -42,6 +42,15 @@
...
 
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
+#include <Kokkos_Macros.hpp>
+#ifndef KOKKOS_ENABLE_DEPRECATED_CODE_3
+static_assert(false,
+              "Including non-public Kokkos header files is not allowed.");
+#else
+KOKKOS_IMPL_WARNING("Including non-public Kokkos header files is not allowed.")
+#endif
+#endif
```
Comment (dalg24) [inline core/src/KokkosExp_MDRangePolicy.hpp:53]: Should be inside the header guards

**Comment by crtrott** [comment]
Where?

**Comment by crtrott** [comment]
We probably should squash merge

**Review by masterleinad — COMMENTED** [review]
I feel uncomfortable with just defining `KOKKOS_IMPL_PUBLIC_INCLUDE` in the tests unconditionally. That's the only way to actually check anything in this pull request.

Also, I think we should check that `KOKKOS_IMPL_PUBLIC_INCLUDE` isn't defined after including any file. I t should be easy enough to add that at least to the headers checked via "core/unit_test/headers_self_contained/tstHeader.cpp"

**Comment by crtrott** [comment]
We still tests this in containers/algorithms and core/perftests (i.e. we do not add the define in those). This also allows is to check inside of core unit tests whether the self sustainability of internal headers is correct.

**Comment by masterleinad** [comment]
The changes I need to allow not adding `KOKKOS_IMPL_PUBLIC_INCLUDE` unconditionally look like
```diff
diff --git a/core/unit_test/CMakeLists.txt b/core/unit_test/CMakeLists.txt
index 4cc31f5fb..ad030285a 100644
--- a/core/unit_test/CMakeLists.txt
+++ b/core/unit_test/CMakeLists.txt
@@ -69,10 +69,6 @@ KOKKOS_INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR})
 KOKKOS_INCLUDE_DIRECTORIES(REQUIRED_DURING_INSTALLATION_TESTING ${CMAKE_CURRENT_SOURCE_DIR})
 KOKKOS_INCLUDE_DIRECTORIES(${KOKKOS_SOURCE_DIR}/core/unit_test/category_files)
 
-# Not pretty but for now best option, since we don't want to add it to all tests (Container/algorithm)
-# and adding it to the almost 50 targets individually in this directory seems also worse
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DKOKKOS_IMPL_PUBLIC_INCLUDE")
-
 SET(COMPILE_ONLY_SOURCES
   TestArray.cpp
   TestDetectionIdiom.cpp
diff --git a/core/unit_test/TestMathematicalFunctions.hpp b/core/unit_test/TestMathematicalFunctions.hpp
index 10ee146ad..a3b055c02 100644
--- a/core/unit_test/TestMathematicalFunctions.hpp
+++ b/core/unit_test/TestMathematicalFunctions.hpp
@@ -48,8 +48,6 @@
 #include <algorithm>
 #include <initializer_list>
 #include <type_traits>
-#include "Kokkos_ExecPolicy.hpp"
-#include "Kokkos_Parallel_Reduce.hpp"
 
 #include <cfloat>
 
diff --git a/core/unit_test/TestNumericTraits.hpp b/core/unit_test/TestNumericTraits.hpp
index e06ac3a46..bb02cbb2a 100644
--- a/core/unit_test/TestNumericTraits.hpp
+++ b/core/unit_test/TestNumericTraits.hpp
@@ -48,7 +48,6 @@
 #include <type_traits>
 #include <limits>
 #include "Kokkos_NumericTraits.hpp"
-#include "Kokkos_ExecPolicy.hpp"
 
 struct extrema {
 #define DEFINE_EXTREMA(T, m, M)                 \
diff --git a/core/unit_test/TestTaskScheduler.hpp b/core/unit_test/TestTaskScheduler.hpp
index 6b9cd2c90..62fd68b6c 100644
--- a/core/unit_test/TestTaskScheduler.hpp
+++ b/core/unit_test/TestTaskScheduler.hpp
@@ -48,7 +48,6 @@
 #include <Kokkos_Macros.hpp>
 #if defined(KOKKOS_ENABLE_TASKDAG)
 #include <Kokkos_Core.hpp>
-#include <impl/Kokkos_FixedBufferMemoryPool.hpp>
 #include <cstdio>
 #include <iostream>
 #include <cmath>
diff --git a/core/unit_test/TestViewHooks.hpp b/core/unit_test/TestViewHooks.hpp
index 1b20e348e..ac679bedb 100644
--- a/core/unit_test/TestViewHooks.hpp
+++ b/core/unit_test/TestViewHooks.hpp
@@ -46,7 +46,6 @@
 
 #include <gtest/gtest.h>
 #include <Kokkos_Core.hpp>
-#include <Kokkos_View.hpp>
 
 namespace Test {
 template <class DeviceType>
diff --git a/core/unit_test/TestViewLayoutTiled.hpp b/core/unit_test/TestViewLayoutTiled.hpp
index 2510a1244..b0c660069 100644
--- a/core/unit_test/TestViewLayoutTiled.hpp
+++ b/core/unit_test/TestViewLayoutTiled.hpp
@@ -42,6 +42,10 @@
 //@HEADER
 */
 
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
+#define KOKKOS_IMPL_PUBLIC_INCLUDE
+#endif
+
 #include <cstdio>
 
 #include <gtest/gtest.h>
@@ -1776,3 +1780,6 @@ TEST(TEST_CATEGORY, view_layouttiled_subtile) {
       4, 12, 16, 12);
 }
 }  // namespace Test
+
+#undef KOKKOS_IMPL_PUBLIC_INCLUDE
+
diff --git a/core/unit_test/headers_self_contained/tstHeader.cpp b/core/unit_test/headers_self_contained/tstHeader.cpp
index d488f0fa3..394a86d8a 100644
--- a/core/unit_test/headers_self_contained/tstHeader.cpp
+++ b/core/unit_test/headers_self_contained/tstHeader.cpp
@@ -4,6 +4,8 @@
 #define KOKKOS_HEADER_TO_TEST \
   KOKKOS_HEADER_TEST_STRINGIZE(KOKKOS_HEADER_TEST_NAME)
 
+#define KOKKOS_IMPL_PUBLIC_INCLUDE 
+
 // include header twice to see if the include guards are set correctly
 #include KOKKOS_HEADER_TO_TEST
 #include KOKKOS_HEADER_TO_TEST
```
So apart from the self-inclusion test, there wasn't much to fix.

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
Retest this please

**Comment by tylerjereddy** [comment]
For reference, I do actually need to include some of these in `pykokkos-base` per https://github.com/kokkos/pykokkos-base/pull/43 I think. Suggestions? I'll ask on Slack as well.

**Comment by dalg24** [comment]
You do not need these.  See kokkos/pykokkos-base#45

**Pull request merged**